desafio-05/ancogamer/go/func.go

```diff
@@ -0,0 +1,364 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+// Area ..
+type Area struct {
+	Codigo string `json:"codigo"`
+	Nome   string `json:"nome"`
+	Qtd    int
+	MediaS float64
+	MaiorS float64
+	MenorS float64
+	//MenorSFuncionarioPointer *Func
+
+	AreaMax string
+	AreaMin string
+}
+
+//Func ..
+type Func struct {
+	ID        int     `json:"id"`
+	Nome      string  `json:"nome"`
+	Sobrenome string  `json:"sobrenome"`
+	Salario   float64 `json:"salario"`
+	Area      string  `json:"area"`
+}
+
+//JSON ..
+type JSON struct {
+	Funcs []Func `json:"funcionarios"`
+	Areas []Area `json:"areas"`
+}
+
+//2
+var (
+	AvgA        string
+	AreasMax    string
+	AreasMin    string
+	c           int
+	LastNameMax string
+)
+
+//SobreS .
+type SobreS struct {
+	LastNameMax  string
+	MaiorS       float64
+	SobreNome    string
+	Funcionarios []*string
+}
+
+//X ..
+func X(bigSalaryByLastName *[]SobreS, dat JSON) {
+	var l int
+	for l = 0; l < len(*bigSalaryByLastName); l++ {
+		if dat.Funcs[c].Sobrenome == (*bigSalaryByLastName)[l].SobreNome {
+			for k := 0; k < len((*bigSalaryByLastName)[l].Funcionarios); k++ {
+				if &dat.Funcs[c].Nome == (*bigSalaryByLastName)[l].Funcionarios[k] {
+					return
+				}
+			}
+			if dat.Funcs[c].Salario > (*bigSalaryByLastName)[l].MaiorS {
+				(*bigSalaryByLastName)[l].MaiorS = dat.Funcs[c].Salario
+				(*bigSalaryByLastName)[l].LastNameMax = ""
+				(*bigSalaryByLastName)[l].LastNameMax = Build((*bigSalaryByLastName)[l].LastNameMax, "\nlast_name_max|", dat.Funcs[c].Sobrenome, "|", dat.Funcs[c].Nome,
+					" ", dat.Funcs[c].Sobrenome, "|",
+					fmt.Sprintf("%.2f", dat.Funcs[c].Salario))
+			} else if dat.Funcs[c].Salario == (*bigSalaryByLastName)[l].MaiorS {
+				(*bigSalaryByLastName)[l].LastNameMax = Build((*bigSalaryByLastName)[l].LastNameMax, "\nlast_name_max|", dat.Funcs[c].Sobrenome, "|", dat.Funcs[c].Nome,
+					" ", dat.Funcs[c].Sobrenome, "|",
+					fmt.Sprintf("%.2f", dat.Funcs[c].Salario))
+			}
+			return
+		}
+	}
+	*bigSalaryByLastName = append(*bigSalaryByLastName, SobreS{
```

Comment:
A lógica está meio difícil de entender pra mim. Grande parte do tempo acima, voce acessa bigSalaryByLastName diretamente via um ponteiro e modifica o elemento "in place". Mas aqui voce apenda algo a ele?

Outra coisa: Se voce for mudar o tamanho de slices, é bom retornar o slice pela função, ao invés de passar um ponteiro. Mas isso depende de como voce implementar a sua função.